File: src/services/export/search/export_search.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type t = {
  index: Export_index.t;
  value_matcher: Fuzzy_path.t;
  type_matcher: Fuzzy_path.t;
}

type search_result = {
  name: string;
  source: Export_index.source;
  kind: Export_index.kind;
}
[@@deriving show]

type search_result_scored = {
  search_result: search_result;
  score: int;
  weight: int;
}
[@@deriving show]

type search_results = {
  results: search_result_scored list;
  is_incomplete: bool;
}
[@@deriving show]

type search_options = Fuzzy_path.options = {
  first_match_can_be_weak: bool;
  num_threads: int;
  max_results: int;
  weighted: bool;
}

let default_options : search_options =
  Fuzzy_path.{ default_options with first_match_can_be_weak = false }

type candidates = {
  values: (string * int) list;
  types: (string * int) list;
}

let summarize_exports exports =
  Export_index.ExportMap.fold
    (fun (_file, kind) num (has_value, has_type, max_count) ->
      let max_count = max num max_count in
      match kind with
      | Export_index.Default -> (true, has_type, max_count)
      | Export_index.Named -> (true, has_type, max_count)
      | Export_index.NamedType -> (has_value, true, max_count)
      | Export_index.Namespace -> (true, has_type, max_count))
    exports
    (false, false, 0)

let partition_candidates index =
  Export_index.fold_names
    ~f:(fun { values; types } name exports ->
      let (has_value, has_type, max_count) = summarize_exports exports in
      let values =
        if has_value then
          (name, max_count) :: values
        else
          values
      in
      let types =
        if has_type then
          (name, max_count) :: types
        else
          types
      in
      { values; types })
    ~init:{ values = []; types = [] }
    index

let init index =
  let { values; types } = partition_candidates index in
  let value_matcher = Fuzzy_path.init values in
  let type_matcher = Fuzzy_path.init types in
  { index; value_matcher; type_matcher }

let subtract old_index { index; value_matcher; type_matcher } =
  let (index, dead_candidates) = Export_index.subtract old_index index in
  let value_matcher = Fuzzy_path.remove_candidates value_matcher dead_candidates in
  let type_matcher = Fuzzy_path.remove_candidates type_matcher dead_candidates in
  { index; value_matcher; type_matcher }

let subtract_count removed_imports { index; value_matcher; type_matcher } =
  let index = Export_index.subtract_count removed_imports index in
  { index; value_matcher; type_matcher }

let merge new_index { index; value_matcher; type_matcher } =
  let index = Export_index.merge new_index index in
  let { values; types } = partition_candidates new_index in
  let value_matcher = Fuzzy_path.add_candidates value_matcher values in
  let type_matcher = Fuzzy_path.add_candidates type_matcher types in
  { index; value_matcher; type_matcher }

(*Merge_import *)
let merge_export_import new_index { index; value_matcher; type_matcher } =
  let index = Export_index.merge_export_import new_index index in
  { index; value_matcher; type_matcher }

type query =
  | Value of string
  | Type of string

let string_of_query = function
  | Value str
  | Type str ->
    str

let search_result_of_export ~query name source kind =
  let open Export_index in
  match (query, kind) with
  | (Value _, (Default | Named | Namespace))
  | (Type _, NamedType) ->
    Some { name; source; kind }
  | (Value _, NamedType)
  | (Type _, (Default | Named | Namespace)) ->
    None

let compare_search_result ~compare_score a b =
  match compare_score a b with
  | 0 ->
    (match Int.compare b.weight a.weight with
    | 0 ->
      (match String.compare a.search_result.name b.search_result.name with
      | 0 ->
        Export_index.compare_export
          (a.search_result.source, a.search_result.kind)
          (b.search_result.source, b.search_result.kind)
      | k -> k)
    | k -> k)
  | k -> k

(** [take ~n:20 ~index matches] will return up to 20 search results,
    where each match in [matches] might contribute multiple results.
    sets [is_incomplete] if [n] is exceeded. *)
let take ~weighted ~n ~index ~query fuzzy_matches =
  let rev_all =
    Base.List.fold fuzzy_matches ~init:[] ~f:(fun acc { Fuzzy_path.value; score } ->
        Export_index.ExportMap.fold
          (fun (source, kind) count acc ->
            match search_result_of_export ~query value source kind with
            | Some search_result ->
              let weight =
                if weighted then
                  count
                else
                  0
              in
              { search_result; score; weight } :: acc
            | None -> acc)
          (Export_index.find value index)
          acc
    )
  in
  let sorted =
    Base.List.stable_sort
      rev_all
      ~compare:
        ((* sorts the highest scores to the front *)
         let compare_score a b = Int.compare b.score a.score in
         fun a b ->
           match compare_search_result ~compare_score a b with
           | 0 ->
             (* since rev_all is reversed, [a; b] where a and b have equal
                scores should return [b; a]. stable_sort wants to leave them
                in the original order. *)
             1
           | k -> k
        )
  in
  let top_n = Base.List.take sorted n in
  let results =
    if weighted then
      (* Exact matches are always sorted first by Fuzzy_path. This results in
         very rare exact matches beating out more common prefix matches.

         For the query `foo`, results `foo` and `foobar` are both contiguous
         prefix matches with the same score, except the exact match gets a +2
         boost.

         We could ignore the boost and order these by weight, but that means
         that exact matches can get buried deep in the list where the user
         will be unlikely to find them. Since they can't keep typing to filter
         (since it's already exact), they're kind of out of luck. So instead,
         we potentially boost the first 3 inexact results by their score
         ignoring the exactness boost, and then by weight.

         For example, suppose the query is `foo` and an autoimport `foo` is
         very rare, `foobar` is very common, and `barfoo` is somewhere in the
         middle. Purely score-wise, it's sorted foo, foobar, barfoo. But we
         remove the boost on foo, so it has the same score as foobar, and then
         re-sort: foobar, foo, barfoo. *)
      let is_exact result =
        String.length result.search_result.name = String.length (string_of_query query)
      in
      let (exacts, rest) = Base.List.split_while top_n ~f:is_exact in
      let (next_3, rest) = Base.List.split_n rest 3 in
      let adjust_score result =
        if is_exact result then
          result.score - 2
        else
          result.score
      in
      let compare_score a b = Int.compare (adjust_score b) (adjust_score a) in
      let top =
        Base.List.stable_sort (exacts @ next_3) ~compare:(compare_search_result ~compare_score)
      in
      top @ rest
    else
      top_n
  in
  let is_incomplete = Base.List.length sorted > n in
  { results; is_incomplete }

let search ?(options = default_options) query { index; value_matcher; type_matcher } =
  let (matcher, query_txt) =
    match query with
    | Value txt -> (value_matcher, txt)
    | Type txt -> (type_matcher, txt)
  in

  let max_results = options.Fuzzy_path.max_results in
  let options =
    (* if max_results is set, then increase it by 1 so that we can tell the difference
       between getting exactly max_results results (is_incomplete = false) and getting
       too many (is_incomplete = true). *)
    if max_results < max_int then
      Fuzzy_path.{ options with max_results = max_results + 1 }
    else
      options
  in

  let weighted = options.Fuzzy_path.weighted in

  Fuzzy_path.search ~options query_txt matcher |> take ~weighted ~n:max_results ~index ~query

let search_values ?options query t = search ?options (Value query) t

let search_types ?options query t = search ?options (Type query) t

let get name { index; value_matcher = _; type_matcher = _ } = Export_index.find name index

let get_values name t =
  get name t
  |> Export_index.ExportMap.filter (fun (_file_key, kind) _num -> Export_index.kind_is_value kind)

let get_types name t =
  get name t
  |> Export_index.ExportMap.filter (fun (_file_key, kind) _num -> Export_index.kind_is_type kind)

let pp fmt { index; _ } = Export_index.pp fmt index

let show { index; _ } = Export_index.show index

File: rust_port/crates/flow_services_export/src/export_search.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::sync::Arc;
use std::sync::Mutex;

use dupe::Dupe;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;

use crate::export_index;
use crate::export_index::Export;
use crate::export_index::ExportIndex;
use crate::export_index::ExportMap;
use crate::export_index::Kind;
use crate::export_index::Source;
use crate::export_search_types::SearchResult;
use crate::export_search_types::SearchResultScored;
use crate::export_search_types::SearchResults;
use crate::fuzzy_path;
use crate::fuzzy_path::FuzzyPath;

#[derive(Debug, Clone)]
pub struct ExportSearch {
    index: Arc<ExportIndex>,
    value_matcher: Arc<Mutex<FuzzyPath>>,
    type_matcher: Arc<Mutex<FuzzyPath>>,
}

pub type SearchOptions = fuzzy_path::Options;

pub fn default_options() -> SearchOptions {
    let mut opts = fuzzy_path::default_options();
    opts.first_match_can_be_weak = false;
    opts
}

struct Candidates {
    values: Vec<(FlowSmolStr, i32)>,
    types: Vec<(FlowSmolStr, i32)>,
}

fn summarize_exports(exports: &ExportMap<i32>) -> (bool, bool, i32) {
    let mut has_value = false;
    let mut has_type = false;
    let mut max_count: i32 = 0;
    for (Export(_file, kind), num) in exports {
        max_count = max_count.max(*num);
        match kind {
            Kind::DefaultType => {
                has_type = true;
            }
            Kind::Default => {
                has_value = true;
            }
            Kind::Named => {
                has_value = true;
            }
            Kind::NamedType => {
                has_type = true;
            }
            Kind::Namespace => {
                has_value = true;
            }
        }
    }
    (has_value, has_type, max_count)
}

fn partition_candidates(index: &ExportIndex) -> Candidates {
    let mut values: Vec<(FlowSmolStr, i32)> = Vec::new();
    let mut types: Vec<(FlowSmolStr, i32)> = Vec::new();
    for (name, exports) in index {
        let (has_value, has_type, max_count) = summarize_exports(exports);
        if has_value {
            values.push((name.dupe(), max_count));
        }
        if has_type {
            types.push((name.dupe(), max_count));
        }
    }
    Candidates { values, types }
}

pub fn init(index: ExportIndex) -> ExportSearch {
    let Candidates { values, types } = partition_candidates(&index);
    let (value_matcher, type_matcher) = FuzzyPath::init_pair_from_arrays(values, types);
    ExportSearch {
        index: Arc::new(index),
        value_matcher: Arc::new(Mutex::new(value_matcher)),
        type_matcher: Arc::new(Mutex::new(type_matcher)),
    }
}

pub fn merge_available_exports(
    old_available_exports_index: &ExportIndex,
    new_available_exports_index: &ExportIndex,
    search: &ExportSearch,
) -> ExportSearch {
    let (addition_index, removal_index) =
        export_index::diff(old_available_exports_index, new_available_exports_index);
    let (index, dead_candidates) = export_index::subtract(&removal_index, search.index.as_ref());
    search
        .value_matcher
        .lock()
        .expect("export search value matcher mutex poisoned")
        .remove_candidates(&dead_candidates);
    search
        .type_matcher
        .lock()
        .expect("export search type matcher mutex poisoned")
        .remove_candidates(&dead_candidates);
    let index = export_index::merge(&addition_index, &index);
    let Candidates { values, types } = partition_candidates(&addition_index);
    search
        .value_matcher
        .lock()
        .expect("export search value matcher mutex poisoned")
        .add_candidates(values);
    search
        .type_matcher
        .lock()
        .expect("export search type matcher mutex poisoned")
        .add_candidates(types);
    ExportSearch {
        index: Arc::new(index),
        value_matcher: search.value_matcher.clone(),
        type_matcher: search.type_matcher.clone(),
    }
}

pub fn subtract_count(removed_imports: &ExportIndex, search: &ExportSearch) -> ExportSearch {
    let index = export_index::subtract_count(removed_imports, search.index.as_ref());
    ExportSearch {
        index: Arc::new(index),
        value_matcher: search.value_matcher.clone(),
        type_matcher: search.type_matcher.clone(),
    }
}

// (*Merge_import *)
pub fn merge_export_import(new_index: &ExportIndex, search: &ExportSearch) -> ExportSearch {
    let index = export_index::merge_export_import(new_index, search.index.as_ref());
    ExportSearch {
        index: Arc::new(index),
        value_matcher: search.value_matcher.clone(),
        type_matcher: search.type_matcher.clone(),
    }
}

enum Query {
    Value(String),
    Type(String),
    BothValueAndType(String),
}

fn string_of_query(query: &Query) -> &str {
    match query {
        Query::Value(s) | Query::Type(s) | Query::BothValueAndType(s) => s,
    }
}

fn search_result_of_export(
    query: &Query,
    name: &str,
    source: Source,
    kind: Kind,
) -> Option<SearchResult> {
    match (query, &kind) {
        (
            Query::Value(_) | Query::BothValueAndType(_),
            Kind::Default | Kind::Named | Kind::Namespace,
        )
        | (Query::Type(_) | Query::BothValueAndType(_), Kind::DefaultType | Kind::NamedType) => {
            Some(SearchResult {
                name: FlowSmolStr::new(name),
                source,
                kind,
            })
        }
        (Query::Value(_), Kind::DefaultType | Kind::NamedType)
        | (Query::Type(_), Kind::Default | Kind::Named | Kind::Namespace) => None,
    }
}

fn compare_search_result(
    compare_score: &impl Fn(&SearchResultScored, &SearchResultScored) -> std::cmp::Ordering,
    a: &SearchResultScored,
    b: &SearchResultScored,
) -> std::cmp::Ordering {
    match compare_score(a, b) {
        std::cmp::Ordering::Equal => match b.weight.cmp(&a.weight) {
            std::cmp::Ordering::Equal => match a.search_result.name.cmp(&b.search_result.name) {
                std::cmp::Ordering::Equal => {
                    let a_export =
                        Export(a.search_result.source.clone(), a.search_result.kind.clone());
                    let b_export =
                        Export(b.search_result.source.clone(), b.search_result.kind.clone());
                    a_export.cmp(&b_export)
                }
                k => k,
            },
            k => k,
        },
        k => k,
    }
}

/// [take ~n:20 ~index matches] will return up to 20 search results,
/// where each match in [matches] might contribute multiple results.
/// sets [is_incomplete] if [n] is exceeded.
fn take(
    weighted: bool,
    n: usize,
    index: &ExportIndex,
    query: &Query,
    fuzzy_matches: &[fuzzy_path::MatchResult],
) -> SearchResults {
    let mut rev_all: Vec<SearchResultScored> = Vec::new();
    for m in fuzzy_matches {
        let exports = export_index::find(&m.value, index);
        for (Export(source, kind), count) in &exports {
            match search_result_of_export(query, &m.value, source.clone(), kind.clone()) {
                Some(search_result) => {
                    let weight = if weighted { *count } else { 0 };
                    rev_all.insert(
                        0,
                        SearchResultScored {
                            search_result,
                            score: m.score,
                            weight,
                        },
                    );
                }
                None => {}
            }
        }
    }

    let compare_score_fn = |a: &SearchResultScored, b: &SearchResultScored| -> std::cmp::Ordering {
        b.score.cmp(&a.score)
    };
    rev_all.sort_by(|a, b| compare_search_result(&compare_score_fn, a, b));
    let sorted = rev_all;

    let top_n: Vec<SearchResultScored> = sorted.iter().take(n).cloned().collect();

    let results = if weighted {
        let query_len = string_of_query(query).len();
        let is_exact =
            |result: &SearchResultScored| -> bool { result.search_result.name.len() == query_len };
        let mut exacts: Vec<SearchResultScored> = Vec::new();
        let mut rest: Vec<SearchResultScored> = Vec::new();
        let mut found_non_exact = false;
        for item in &top_n {
            if !found_non_exact && is_exact(item) {
                exacts.push(item.clone());
            } else {
                found_non_exact = true;
                rest.push(item.clone());
            }
        }
        let next_3: Vec<SearchResultScored> = rest.iter().take(3).cloned().collect();
        let rest: Vec<SearchResultScored> = rest.into_iter().skip(3).collect();
        let adjust_score = |result: &SearchResultScored| -> i32 {
            if is_exact(result) {
                result.score - 2
            } else {
                result.score
            }
        };
        let compare_score_adjusted =
            |a: &SearchResultScored, b: &SearchResultScored| -> std::cmp::Ordering {
                adjust_score(b).cmp(&adjust_score(a))
            };
        let mut top: Vec<SearchResultScored> = exacts;
        top.extend(next_3);
        top.sort_by(|a, b| compare_search_result(&compare_score_adjusted, a, b));
        top.extend(rest);
        top
    } else {
        top_n
    };

    let is_incomplete = sorted.len() > n;
    SearchResults {
        results,
        is_incomplete,
    }
}

fn search_internal(options: &SearchOptions, query: Query, search: &ExportSearch) -> SearchResults {
    let max_results = options.max_results;
    let search_options = if max_results < usize::MAX {
        let mut opts = options.clone();
        opts.max_results = max_results + 1;
        opts
    } else {
        options.clone()
    };
    let weighted = search_options.weighted;
    let fuzzy_matches = match &query {
        Query::Value(txt) => search
            .value_matcher
            .lock()
            .expect("export search value matcher mutex poisoned")
            .search(txt, &search_options),
        Query::Type(txt) => search
            .type_matcher
            .lock()
            .expect("export search type matcher mutex poisoned")
            .search(txt, &search_options),
        Query::BothValueAndType(_) => {
            panic!("BothValueAndType query should be handled by search_both_values_and_types")
        }
    };
    take(
        weighted,
        max_results,
        search.index.as_ref(),
        &query,
        &fuzzy_matches,
    )
}

pub fn search_values(
    options: Option<&SearchOptions>,
    query: &str,
    t: &ExportSearch,
) -> SearchResults {
    let opts = match options {
        Some(o) => o.clone(),
        None => default_options(),
    };
    search_internal(&opts, Query::Value(query.to_string()), t)
}

pub fn search_types(
    options: Option<&SearchOptions>,
    query: &str,
    t: &ExportSearch,
) -> SearchResults {
    let opts = match options {
        Some(o) => o.clone(),
        None => default_options(),
    };
    search_internal(&opts, Query::Type(query.to_string()), t)
}

pub fn search_both_values_and_types(
    options: Option<&SearchOptions>,
    query: &str,
    t: &ExportSearch,
) -> SearchResults {
    let options = match options {
        Some(o) => o.clone(),
        None => default_options(),
    };
    let max_results = options.max_results;
    let search_options = if max_results < usize::MAX {
        let mut opts = options.clone();
        opts.max_results = max_results + 1;
        opts
    } else {
        options.clone()
    };

    let weighted = search_options.weighted;
    let value_fuzzy_results = t
        .value_matcher
        .lock()
        .expect("export search value matcher mutex poisoned")
        .search(query, &search_options);
    let type_fuzzy_results = t
        .type_matcher
        .lock()
        .expect("export search type matcher mutex poisoned")
        .search(query, &search_options);

    let mut combined = value_fuzzy_results;
    combined.extend(type_fuzzy_results);
    let SearchResults {
        results,
        is_incomplete,
    } = take(
        weighted,
        max_results,
        t.index.as_ref(),
        &Query::BothValueAndType(query.to_string()),
        &combined,
    );

    let mut seen: BTreeMap<Export, std::collections::BTreeSet<FlowSmolStr>> = BTreeMap::new();
    let results = results
        .into_iter()
        .filter(|item| {
            let name = &item.search_result.name;
            let source = &item.search_result.source;
            let key = Export(source.clone(), Kind::Named);
            let name_set = seen.entry(key).or_default();
            if name_set.contains(name) {
                false
            } else {
                name_set.insert(name.dupe());
                true
            }
        })
        .collect();

    SearchResults {
        results,
        is_incomplete,
    }
}

// let get_index t = t.index
pub fn get_index(t: &ExportSearch) -> &ExportIndex {
    t.index.as_ref()
}

pub fn get(name: &str, t: &ExportSearch) -> ExportMap<i32> {
    export_index::find(name, t.index.as_ref())
}

pub fn get_values(name: &str, t: &ExportSearch) -> ExportMap<i32> {
    get(name, t)
        .into_iter()
        .filter(|(Export(_file_key, kind), _num)| export_index::kind_is_value(kind))
        .collect()
}

pub fn get_types(name: &str, t: &ExportSearch) -> ExportMap<i32> {
    get(name, t)
        .into_iter()
        .filter(|(Export(_file_key, kind), _num)| export_index::kind_is_type(kind))
        .collect()
}

impl std::fmt::Display for ExportSearch {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self.index)
    }
}
